src/common/libschedutil/init.c

```diff
@@ -0,0 +1,111 @@
+/************************************************************\
+ * Copyright 2019 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+#include <czmq.h>
+#include <errno.h>
+#include <flux/core.h>
+
+#include "src/common/libutil/aux.h"
+
+#include "init.h"
+
+struct schedutil_ctx {
+    flux_t *h;
+    struct aux_item *aux;
+    bool test_pause; // used for testing that schedutil responds to outstanding futures when sched is unloaded
+    zlistx_t *outstanding_msgs;
+};
+
+schedutil_t *schedutil_create (flux_t *h)
+{
+    schedutil_t *s = calloc(1, sizeof(*s));
+    if (!s)
+        return NULL;
+
+    s->h = h;
+    s->test_pause = false;
+    s->outstanding_msgs = zlistx_new ();
+
+    return s;
+}
+
+static void respond_to_outstanding_msgs (schedutil_t *s)
+{
+    int rc = 0;
+    flux_msg_t *msg;
+    for (msg = zlistx_first (s->outstanding_msgs);
+         msg;
+         msg = zlistx_next (s->outstanding_msgs))
+        {
+            rc += flux_respond_error (s->h, msg, ENOSYS, "automatic ENOSYS response from schedutil");
+        }
+    if (rc != 0)
+        flux_log(s->h, LOG_ERR, "schedutil: error in responding to outstanding messages");
+    zlistx_purge (s->outstanding_msgs);
+}
+
+void schedutil_destroy (schedutil_t *s)
+{
+    if (s) {
+        int saved_errno = errno;
+        respond_to_outstanding_msgs (s);
+        aux_destroy (&(s->aux));
+        zlistx_destroy (&s->outstanding_msgs);
+        free (s);
+        errno = saved_errno;
+    }
+    return;
+}
+
+flux_t *schedutil_get_flux (const schedutil_t *s)
+{
+    return s->h;
+}
+
+void *schedutil_aux_get (schedutil_t *s, const char *name)
+{
+    if (!s) {
+        errno = EINVAL;
+        return NULL;
+    }
+    return aux_get (s->aux, name);
+}
+
+int schedutil_aux_set (schedutil_t *s, const char *name,
+                        void *aux, flux_free_f destroy)
+{
+    if (!s) {
+        errno = EINVAL;
+        return -1;
+    }
+    return aux_set (&s->aux, name, aux, destroy);
+}
+
+void schedutil_pause (schedutil_t *s)
+{
+    s->test_pause = true;
+}
+
+bool schedutil_is_paused (const schedutil_t *s)
+{
+    return s->test_pause;
+}
+
+int schedutil_add_outstanding_msg (schedutil_t *s, const flux_msg_t *msg)
+{
+    zlistx_add_end (s->outstanding_msgs, (void*)msg);
+    return 0;
+}
+
+int schedutil_remove_outstanding_msg (schedutil_t *s, const flux_msg_t *msg)
+{
+    zlistx_detach (s->outstanding_msgs, (void*)msg);
```

Comment:
I *think* `zlistx_detach` requires a "handle" not the item. So I think you need something like:
```C
void *x = zlistx_find (s->outstanding_msgs, (void *) msg);
if (!x) {
    errno = ENOENT;
    return -1;
}
if (x)
    zlistx_detach (s->outstanding_msgs, x);
```

You might also be able to use `zlistx_detach_cur()`  instead of saving the result of `zlistx_find`

Reply:
Good catch.  Thanks!

Reply:
Fixed in https://github.com/flux-framework/flux-core/pull/2226/commits/797e64e152b2ad3a03b3978be26e23b0f88df6be